Improved Command Line with CLI11 (#179)
BlueBrain/CoreNeuron: README.md

## RUNNING SIMULATION:

Note that the CoreNEURON simulator depends on NEURON to build the network model: see [NEURON](https://www.neuron.yale.edu/neuron/) documentation for more information. Once you build the model using NEURON, you can launch CoreNEURON on the same or different machine by:

```bash
export OMP_NUM_THREADS=2     #set appropriate value
mpiexec -np 2 build/apps/coreneuron_exec -e 10 --mpi input -d /path/to/model/built/by/neuron
```

[This tutorial](https://github.com/nrnhines/ringtest) provide more information for parallel runs and performance comparison.

### Command Line Interface

:warning: :warning: :warning: **In a recent update the command line interface was updated, so please update your scripts accordingly!**

Some details on the new interface:

The new command line interface is based on CLI11. All the previous options are still supported but they are organized differently. You can find more details by running `coreneuron_exec --help-all` or using the table reported below.

> Multiple character options with single dash (e.g. `-gpu`) are **not** supported anymore. All multiple characters options now require a double dash (e.g. `--gpu`), but single characters option still support a single dash (e.g. `-g`).

Reviewer: ```suggestion
Multiple character options with single dash (e.g. `-gpu`) are **not** supported anymore. All multiple characters options now require a double dash (e.g. `--gpu`), but single characters options still support a single dash (e.g. `-g`).
```

> The structure of a command is the following: `./apps/coreneuron_exec [OPTIONS] [SUBCOMMAND]`. More specifically, an `OPTION` is a parameter that doesn't belong to **any** `SUBCOMMANDS`. The `OPTIONS` should **always** be written before any of the `SUBCOMMAND`. To know which of the parameter is an `OPTION` or a `SUBCOMMAND` run `coreneuron_exec --help-all` or refer to the table below.

Reviewer: ```suggestion
The structure of a command is the following: `./apps/coreneuron_exec [GENERIC OPTIONS] [SUBCOMMAND [OPTIONS]]`. More specifically, a `GENERIC OPTION` is a parameter that doesn't belong to **any** `SUBCOMMANDS`. `GENERIC OPTIONS` should **always** be written before any `SUBCOMMAND`. To know which of the various subcommands and their respective options run `coreneuron_exec --help-all` or refer to the table below.
```

> To access a parameter that falls under any of the subcommands category it is necessary to include in the command line a certain keyword first. For example to use the `-d` option, one must enter in the `input` subcommands first by using the `input` keyword:

Reviewer: ```suggestion
To access a subcommand-specific parameter it must always follow that subcommand. For example to use the `-d` option, one must first write the `input` subcommand.
```

> therefore the old command `coreneuron_exec -d /path/to/model/built/by/neuron` becomes `coreneuron_exec input -d /path/to/model/built/by/neuron` in the new implementation. If you want to  define **more than one** parameter in a certain `SUBCOMMAND` it is needed to state the **subcommand keyword** (for example `input`) **directly followed** by the **options** of this subcommand. The `SUBCOMMAND` keyword needs to be stated only once for every set of parameters.

Reviewer: ```suggestion
Hence, the old command `coreneuron_exec -d /path/to/model/built/by/neuron` becomes `coreneuron_exec input -d /path/to/model/built/by/neuron` in the new implementation. All subcommand-specific options must be grouped with their respective subcommand.
```

Note that the CoreNEURON simulator depends on NEURON to build the network model: see [NEURON](https://www.neuron.yale.edu/neuron/) documentation for more information. Once you build the model using NEURON, you can launch CoreNEURON on the same or different machine by:

```bash
export OMP_NUM_THREADS=2     #set appropriate value
mpiexec -np 2 build/apps/coreneuron_exec -e 10 --mpi input -d /path/to/model/built/by/neuron
```

[This tutorial](https://github.com/nrnhines/ringtest) provide more information for parallel runs and performance comparison.

### Command Line Interface

:warning: :warning: :warning: **In a recent update the command line interface was updated, so please update your scripts accordingly!**

Some details on the new interface:

The new command line interface is based on CLI11. All the previous options are still supported but they are organized differently. You can find more details by running `coreneuron_exec --help-all`.

Multiple character options with single dash (e.g. `-gpu`) are not supported anymore. All multiple characters options now require a double dash (e.g. `--gpu`), but single characters option still support a single dash (e.g. `-g`).

> The structure of a command is the following: `./apps/coreneuron_exec [OPTIONS] [SUBCOMMAND]`. The `OPTIONS` should **always** be written before any of the `SUBCOMMAND`. To know which of the parameter is an `OPTION` or a `SUBCOMMAND` run `coreneuron_exec --help-all`.

Reviewer: Do you mean by `OPTIONS` here general options that don't belong to any subcommand? This should be clarified. The current text seems to contradict below paragraph.

Author: That's what I mean! It's pretty unclear so I'll clarify this...

> To access a parameter that falls under any of the subcommands category it is necessary to include in the command line a certain keyworld first. For example to use the `-d` option, one must enter in the `input` subcommands first by using the `input` keyword first:

Reviewer: ```suggestion
To access a parameter that falls under any of the subcommands category it is necessary to include in the command line a certain keyword first. For example to use the `-d` option, one must enter in the `input` subcommands first by using the `input` keyword first:
```

Reviewer: Also, the paragraph should be updated for clarity. It's not really clear, in this form.

Author: By including the definition of `options` as parameters that do not belong to any of the `subcommands` it should be a bit clearer. But I understand it can be a bit foggy. What parts in particular would you improve? What things are not clear for you?

`coreneuron_exec -d /path/to/model/built/by/neuron` becomes `coreneuron_exec input -d /path/to/model/built/by/neuron`

> To define **more than one** option in a certain subcommand it is needed to state the **subcommand keyword** (for example `input`) **directly followed** by the **options** of this subcommand (for example `-d /path/to/model -f /path/to/filesdat/file.dat`). The full command would be:

Reviewer: You seem to use _option_ and _parameter_ interchangeably. This is confusing. Also this paragraph and the previous can be written shorter and as one.

Author: I tried to use the word `parameter` in a generic way, while using `option` for general options that do not belong to any subcommands. In a few places it was inconsistent so I fixed it.

Multiple characters options with single dash (`-gpu`, `-mpi`, `-dt`) are **not** supported anymore. All those options now require a double dash (`--gpu`, `--mpi`, `--dt`), but single characters options still support a single dash (e.g. `-g`).

The structure of a command is the following: `./apps/coreneuron_exec [GENERIC OPTIONS] [SUBCOMMAND [OPTIONS]]`. More specifically, a `GENERIC OPTION` is a parameter that doesn't belong to **any** `SUBCOMMANDS`. `GENERIC OPTIONS` should **always** be written before any `SUBCOMMAND`. To know the various subcommands and their respective options run `coreneuron_exec --help-all` or refer to the table below.

To access a subcommand-specific parameter it must always follow that subcommand. For example to use the `-d` option, one must first write the `input` subcommand.
Hence, the old command `coreneuron_exec -d /path/to/model/built/by/neuron` becomes `coreneuron_exec input -d /path/to/model/built/by/neuron` in the new implementation. All subcommand-specific options must be grouped with their respective subcommand.

In order to see the command line options, you can use:

```bash

> /path/to/install/directory/coreneuron_exec --help-all

Reviewer 2: I plan to remove the print of the help in README?

Any thoughts?

Author: Nothing against it here, actually it makes sense!

[This tutorial](https://github.com/nrnhines/ringtest) provide more information for parallel runs and performance comparison.

###Command Line Interface

:warning: :warning: :warning: **In a recent update the command line interface was updated, so please update your scripts accordingly if necessary!**

Some details on the new interface:

The new command line interface is based on CLI11. All the previous options are still supported but they are organized in subcommands. You can find more details by running `coreneuron_exec --help-all`.

Also multiple character options with single dash (e.g. `-gpu`) are not supported anymore. All multiple characters options require a double dash now (e.g. `--gpu`).

> Since commands are now organized in subcommands, to access a certain parameter it might be necessary to include in the command line a certain keyworld first.

Reviewer 3: I feel that it should be more helpful to state clearer that the `options` should always be before the subcommands

For example to use the `-d` option, one must enter in the `input` subcommands first by using the `input` keyword first:
`coreneuron_exec -d /path/to/model/built/by/neuron` becomes `coreneuron_exec input -d /path/to/model/built/by/neuron`

> Multiple options in a certain subcommand section can be specifying the keyword only once: `coreneuron_exec --mpi input -d /path/to/model -f /path/to/filesdat/file.dat`

Reviewer 3: Maybe this could be rephrased a bit?
```suggestion
To define **more than one** option in a certain subcommand it is needed to state the **subcommand keyword** (for example `input`) **directly followed** by the **options** of this subcommand (for example `-d /path/to/model -f /path/to/filesdat/file.dat`). The full command would be:
`coreneuron_exec --mpi input -d /path/to/model -f /path/to/filesdat/file.dat`
```
Or something similar? Feel free to modify my suggestion if you want
